Introduce CVA to html-extra (#4006)
twigphp/Twig: doc/functions/cva.rst

``cva``
=======

`CVA (Class Variant Authority)`_ is a concept from the JavaScript world and used
by the well-known `shadcn/ui`_.
CVA allows you to display a component with different variants (color, size, etc.),
to create highly reusable and customizable templates. This function work best by being combined with

> `TwigComponent`_

Reviewer: I'd maybe mention this later in the page, as TwigComponent is --despite its name-- a Symfony package.

.. note::

    The ``cva`` function is part of the ``HtmlExtension`` which is not
    installed by default. Install it first:

    .. code-block:: bash

        $ composer require twig/html-extra

>     Then, on Symfony projects, install the ``twig/extra-bundle``:

Reviewer: Similar notes are currently located at the end of the doc pages

* https://twig.symfony.com/doc/3.x/filters/markdown_to_html.html
* https://twig.symfony.com/doc/3.x/functions/timezone_names.html
* https://twig.symfony.com/doc/3.x/functions/html_classes.html

Reviewer: 👀

This is powered by a ``cva()`` Twig

> function where you define ``base`` classes that should always be present and then different

Reviewer: If they must be present, first argument of the function / CVA should not be nullable

Author: This is the classes inside the base that should always be present not the base argument

Reviewer: Maybe we could avoid using "base" for two distinct things then ?

``variants`` and the corresponding classes:

.. code-block:: html+twig

    {# templates/alert.html.twig #}

>      {% set alert = cva({

Reviewer: ```suggestion
      {% set alert = cva({
```

            size: {
                sm: 'text-sm',
                md: 'text-md',
                lg: 'text-lg',
            }
        }
    }) %}

    <div class="{{ alert.apply({color, size}, class)) }}">
        ...
    </div>

Then use the ``color`` and ``size`` variants to select the classes needed:

.. code-block:: twig

    {# index.html.twig #}
    {% include 'alert.html.twig' with {'color': 'blue', 'size': 'md'} %}
    // class="alert bg-red text-lg"

    {% include 'alert.html.twig' with {'color': 'green', 'size': 'sm'} %}
    // class="alert bg-green text-sm"

    {% include 'alert.html.twig' with {'color': 'red', 'class': 'flex items-center justify-center'} %}
    // class="alert bg-red text-md flex items-center justify-center"

CVA and Tailwind CSS
~~~~~~~~~~~~~~~~~~~~

CVA work perfectly with Tailwind CSS. The only drawback is that you can have class conflicts.
To "merge" conflicting classes together and keep only the ones you need, use the
``tailwind_merge()`` filter from `tales-from-a-dev/twig-tailwind-extra`_
with the ``cva()`` function:

.. code-block:: terminal

    $ composer require tales-from-a-dev/twig-tailwind-extra

.. code-block:: html+twig

>    {% set alert = cva({

Reviewer: ```suggestion
    {% set alert = cva({
```

         {% block content %}{% endblock %}
    </div>

    {# index.html.twig #}

    {% include 'alert.html.twig' with {color: 'red', size: 'lg'} %}
    // class="alert bg-red text-lg font-bold"

    {% include 'alert.html.twig' with {color: 'green', size: 'sm'} %}
    // class="alert bg-green text-sm"

    {% include 'alert.html.twig' with {color: 'red', size: 'md'} %}
    // class="alert bg-green text-lg font-bold"

Default Variants
~~~~~~~~~~~~~~~~

If no variants match, you can define a default set of classes to apply:

.. code-block:: html+twig

    {% set alert = cva({
        base: 'alert ',
        variants: {
            color: {
                blue: 'bg-blue',
                red: 'bg-red',
                green: 'bg-green',
            },
            size: {
                sm: 'text-sm',
                md: 'text-md',
                lg: 'text-lg',
            },
            rounded: {
                sm: 'rounded-sm',
                md: 'rounded-md',
                lg: 'rounded-lg',
            }
        },
        defaultVariants: {
            rounded: 'md',
        }
    }) %}

    <div class="{{ alert.apply({color, size}) }}">

>          {% block content %}{% endblock %}

Reviewer: Maybe remove the block here (or use an embed as example ?)

``cva``

> ================

Reviewer 2: ```suggestion
=======
```

`CVA (Class Variant Authority)`_ is a concept from the JavaScript world and used
by the well-known `shadcn/ui`_.
CVA allows you to display a component with different variants (color, size, etc.),

> to create highly reusable and customizable templates. This function work best by being combine with

Reviewer 2: ```suggestion
to create highly reusable and customizable templates. This function work best by being combined with
```

`TwigComponent`_

.. note::

>     The ``html_classes`` function is part of the ``HtmlExtension`` which is not

Reviewer 2: ```suggestion
    The ``cva`` function is part of the ``HtmlExtension`` which is not
```

    installed by default. Install it first:

    .. code-block:: bash

        $ composer require twig/html-extra

    Then, on Symfony projects, install the ``twig/extra-bundle``:

This is powered by a ``cva()`` Twig
function where you define ``base`` classes that should always be present and then different
``variants`` and the corresponding classes:

.. code-block:: html+twig

    {# templates/alert.html.twig #}
     {% set alert = cva({
        base: 'alert ',
        variants: {
            color: {
                blue: 'bg-blue',
                red: 'bg-red',
                green: 'bg-green',
            },
            size: {
                sm: 'text-sm',
                md: 'text-md',
                lg: 'text-lg',
            }
        }
    }) %}

>     <div class="{{ alert.apply({color, size}, attributes.render('class')) }}">

Reviewer 2: can we create an example that does not rely on the special `attributes` variable of TwigComponent ? If this gets merged in Twig itself instead of staying in symfony/ux, it must be applicable to standalone usages.

        ...
    </div>

Then use the ``color`` and ``size`` variants to select the classes needed:

.. code-block:: twig

    {# index.html.twig #}
    {% include 'alert.html.twig' with {'color': 'blue', 'size': 'md'} %}
    // class="alert bg-red text-lg"

    {% include 'alert.html.twig' with {'color': 'green', 'size': 'sm'} %}
    // class="alert bg-green text-sm"

    {% include 'alert.html.twig' with {'color': 'red', 'class': 'flex items-center justify-center'} %}
    // class="alert bg-red text-md flex items-center justify-center"

CVA and Tailwind CSS
~~~~~~~~~~~~~~~~~~~~

CVA work perfectly with Tailwind CSS. The only drawback is that you can have class conflicts.
To "merge" conflicting classes together and keep only the ones you need, use the

> ``tailwind_merge()`` method from `tales-from-a-dev/twig-tailwind-extra`_

Reviewer 2: this is a filter, not a method (Twig does not have methods)
